shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Timelock.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE DataKinds      #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE FlexibleInstances #-}
+
+module Shelley.Spec.Ledger.Timelock
+  ( Timelock( Interval, TimelockAnd, TimelockOr),
+    ininterval
+  )
+  where
+
+import Cardano.Slotting.Slot ( SlotNo (..))
+import Cardano.Binary
+  ( Annotator (..),
+    FromCBOR (fromCBOR),
+    ToCBOR (toCBOR),
+    annotatorSlice,
+    encodeListLen,
+    encodePreEncoded,
+    encodeWord,
+    serializeEncoding,
+  )
+import Cardano.Ledger.Era(Era)
+import qualified Data.ByteString.Lazy as BSL
+import GHC.Generics (Generic)
+import NoThunks.Class (AllowThunksIn (..), NoThunks (..))
+import Shelley.Spec.Ledger.BaseTypes (invalidKey)
+import Shelley.Spec.Ledger.Serialization (decodeList, decodeRecordSum, encodeFoldable)
+
+
+-- ================================================================
+-- Timelock' and Timelock are mutually recursive. Timelock adds a
+-- convenient place to hang the memoized serialized bytes.
+
+data Timelock' era
+   = Interval' !(Maybe SlotNo,Maybe SlotNo)
+   | TimelockAnd' ![ Timelock era ]    -- Note the hidden recursion of Timelock', through Timelock.
+   | TimelockOr'  ![ Timelock era ]
```

Comment:
Note that these bangs are not enough to prevent any thunks:

* The tuple is lazy -> use two different arguments with bangs to the constructors
* `Maybe` is lazy -> use `StrictMaybe`
* A list is list -> use `StrictSeq`

Reply:
the MA spec has another constructor for multisig scripts

Reply:
Fixed all this